### DISCUSSION TITLE
"ActiveDocument" discussion

### DISCUSSION THREAD

**Discussion by johnnyshields** [Dynamoid/dynamoid#741] [General]
As proposed in [this issue](https://github.com/Dynamoid/dynamoid/issues/739), I'd like to start a discussion on building "ActiveDocument", which would be a database-agnostic Ruby ODM framework for document-based databases analogous to ActiveRecord. Target databases would include:
- MongoDB
- DocumentDB
- CouchDB
- Possibly Elasticsearch, Postgres JSONB, etc.

Core ideas:
- Ideally functions as a drop-in replacement for both Dynamoid and Mongoid.
- Standardize where it makes sense to standardize, but preserve as much DB-specific functionality as possible.
- Fully embrace the document concept. Make it as easy/efficient as possible to work with documents in Ruby memory, with high performance for large workloads.
- An `ActiveDocument::Document` class will model fields and relations in a DB-agnostic manner. Fields will have a type definition, and embedded documents will also be supported.
- Add query adapters for each DB. Common methods like `.where`, `.gt`, `.lt` etc. will be supported, but there will also be language-specific query operators as well such as `.where(name: { "$in" => ["Bob", "Jim"])` for MongoDB.
- There would also be classes to handle type coercion to and from Ruby, with slight differences for serialization to DB, deserialization from DB, and serialization for query commands.

I'm a frequent contributor to Mongoid (we use it at [TableCheck](https://www.tablecheck.com/en/join/)) which Dynamoid was originally forked from. Starting the project as a merger between Mongoid and Dynamoid might be a good entry point.

Thoughts?

**Comment by johnnyshields** [comment]
@andrykonchin @pboling update on this--I've created [activedocument/activedocument repo](https://github.com/activedocument/activedocument) and I've also secured the gem ownership for [activedocument](https://rubygems.org/gems/activedocument).

Mongoid is far more full-featured than Dynamoid so I will be using it as a base. On the flip-side, much of Mongoid is bloated and needlessly complex, so I plan to refactor much of it but keep all the test-cases.

[Roadmap is here](https://github.com/activedocument/activedocument/issues/1), please let me know if you're interested to help out.

**Reply by pboling** [reply]
I love the idea!  I would love to consolidate efforts on a best-in-class solution with broad support, akin to ActiveRecord, or DrizzleORM in the JS world.

I would love to help out!  In my day job we are, as I understand it, one of the, or the, largest enterprise client of MongoDB's cloud offering, so that might be an interesting use case.

I do not have a current use case for DynamoDB.  Happy to be added to the org/project however you see fit!

**Reply by johnnyshields** [reply]
@pboling i've invited you to the project!

**Reply by pboling** [reply]
Let's GOOOOO!

**Reply by ckhsponge** [reply]
@johnnyshields I added transaction support for Dynamoid and would be willing to help with DynamoDB transactions for ActiveDocument.

**Reply by johnnyshields** [reply]
@ckhsponge very cool! ActiveDocument (derived from Mongoid) actually supports transactions today, at least for MongoDB.

I think a first order of business is to add DynamoDB support to ActiveDocument, which isn't done yet. We're currently working on a [database-agnostic query AST](https://github.com/activedocument/activedocument/issues/15) which will handle the query side of the equation. I would very much welcome someone from Dynamoid to start a PR on the persistence layer, which is probably largely about porting over Dynamoid code. Importantly we'll have to separate a database-agnostic "core" (field definitions, relation definitions, type-casting, etc.) from the actual persistence API calls. At the moment all of this stuff is much more coupled to MongoDB than it should be.
